pennylane_lightning/core/src/simulators/lightning_gpu/gates/cuGateCache.hpp

```diff
@@ -0,0 +1,281 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cmath>
+#include <complex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "DataBuffer.hpp"
+#include "DevTag.hpp"
+#include "Gates.hpp"
```

Comment:
Doesn't the complex matrix datatype differ between them though?

Reply:
Yes, my idea is have all definitions in `Gates.hpp` and a function (e.g. `vector_complex_2_cucomplex`) that converts the gates' output to the correct type for backends that require it. 

Actually, now that I'm looking at `Gates.hpp`, the functions are templated on the complex type as follows
```
template <template <typename...> class ComplexT, typename T>
```
I think did that to accommodate `Kokkos::complex`. So can't we use CUDA complex types here?

Reply:
It doesn't support with our current design. It could be fixed if we add an abstract `ComplexT<T>` to wrap to `cuFloatComplex` and `cuDoubleComplex`. I guess this can be also a way to let us deprecate `CFP_t`. Let's do it in a follow-up PR.